velox/common/memory/Memory.h

```diff
@@ -65,89 +65,174 @@ constexpr uint16_t kDefaultAlignment = 64;
-  // propagates down the subtree recursively.
+  /// Called by MemoryManager and MemoryPool upon memory usage updates and
+  /// propagates down the subtree recursively.
   virtual void capMemoryAllocation() = 0;
-  // Called by MemoryManager and propagates down recursively if applicable.
+  /// Called by MemoryManager and propagates down recursively if applicable.
   virtual void uncapMemoryAllocation() = 0;
-  // We might need to freeze memory allocating operations under severe global
-  // memory pressure.
+  /// We might need to freeze memory allocating operations under severe global
+  /// memory pressure.
   virtual bool isMemoryCapped() const = 0;
 
-  ////////////////////////// tree methods ////////////////////////////
-  // TODO: might be able to move implementation to abstract class
-  // name of the node, preferably unique among siblings, but not enforced.
-  virtual const std::string& getName() const = 0;
-  virtual std::weak_ptr<MemoryPool> getWeakPtr() = 0;
-  virtual MemoryPool& getChildByName(const std::string& name) = 0;
-  virtual uint64_t getChildCount() const = 0;
-  virtual void visitChildren(
-      std::function<void(MemoryPool* FOLLY_NONNULL)> visitor) const = 0;
-
+ protected:
+  /// Invoked by addChild() to create a child memory pool object. 'parent' is
+  /// a shared pointer created from this.
   virtual std::shared_ptr<MemoryPool> genChild(
-      std::weak_ptr<MemoryPool> parent,
+      std::shared_ptr<MemoryPool> parent,
       const std::string& name,
       int64_t cap) = 0;
-  virtual MemoryPool& addChild(
-      const std::string& name,
-      int64_t cap = kMaxMemory) = 0;
-  virtual std::unique_ptr<ScopedMemoryPool> addScopedChild(
-      const std::string& name,
-      int64_t cap = kMaxMemory) = 0;
-  virtual void dropChild(const MemoryPool* FOLLY_NONNULL child) = 0;
 
-  // Used to explicitly remove self when a user is done with the node, since
-  // nodes are owned by parents and users only get object references. Any
-  // reference to self would be invalid after this call. This will also drop the
-  // entire subtree.
-  virtual void removeSelf() = 0;
+  /// Invoked only on object destructor to remove this memory pool from its
+  /// parent's child memory pool tracking.
+  virtual void dropChild(const MemoryPool* FOLLY_NONNULL child);
 
-  // Used to manage existing externally allocated memories without doing a new
-  // allocation.
-  virtual void reserve(int64_t /* bytes */) {
-    VELOX_NYI("reserve() needs to be implemented in derived memory pool.");
-  }
-  // Sometimes in memory governance we want to mock an update for quota
-  // accounting purposes and different implementations can
-  // choose to accommodate this differently.
-  virtual void release(int64_t /* bytes */, bool /* mock */ = false) {
-    VELOX_NYI("release() needs to be implemented in derived memory pool.");
-  }
+  const std::string name_;
+  std::shared_ptr<MemoryPool> parent_;
+
+  mutable folly::SharedMutex childrenMutex_;
```

Comment:
nit: perhaps, add a short comment to explain that this is used to process access to 'children_'